packages/attribute-aware-php-doc/src/Ast/Type/AttributeAwareArrayTypeNode.php

```diff
@@ -6,26 +6,72 @@
 
 use PHPStan\PhpDocParser\Ast\Type\ArrayTypeNode;
 use PHPStan\PhpDocParser\Ast\Type\CallableTypeNode;
+use PHPStan\PhpDocParser\Ast\Type\TypeNode;
 use PHPStan\PhpDocParser\Ast\Type\UnionTypeNode;
 use Rector\BetterPhpDocParser\Attributes\Attribute\AttributeTrait;
 use Rector\BetterPhpDocParser\Contract\PhpDocNode\AttributeAwareNodeInterface;
+use Rector\PHPStanStaticTypeMapper\TypeMapper\ArrayTypeMapper;
 
 final class AttributeAwareArrayTypeNode extends ArrayTypeNode implements AttributeAwareNodeInterface
 {
     use AttributeTrait;
 
     public function __toString(): string
     {
-        $typeAsString = (string) $this->type;
-
         if ($this->type instanceof CallableTypeNode) {
-            return sprintf('(%s)[]', $typeAsString);
+            return sprintf('(%s)[]', (string) $this->type);
         }
 
-        if ($this->type instanceof UnionTypeNode || $this->type instanceof ArrayTypeNode) {
+        $typeAsString = (string) $this->type;
+
+        if ($this->isGenericArrayCandidate($this->type)) {
             return sprintf('array<%s>', $typeAsString);
         }
 
+        if ($this->type instanceof ArrayTypeNode) {
+            return $this->printArrayType($this->type);
+        }
+
+        if ($this->type instanceof AttributeAwareUnionTypeNode) {
+            return $this->printUnionType($this->type);
+        }
+
         return $typeAsString . '[]';
     }
+
+    private function isGenericArrayCandidate(TypeNode $typeNode): bool
+    {
+        if (! $this->getAttribute(ArrayTypeMapper::HAS_GENERIC_TYPE_PARENT)) {
+            return false;
+        }
+
+        return $typeNode instanceof UnionTypeNode || $typeNode instanceof ArrayTypeNode;
+    }
+
+    private function printArrayType(ArrayTypeNode $arrayTypeNode): string
+    {
+        $typeAsString = (string) $arrayTypeNode;
+
+        $singleTypesAsString = explode('|', $typeAsString);
+        foreach ($singleTypesAsString as $key => $singleTypeAsString) {
+            $singleTypesAsString[$key] = $singleTypeAsString . '[]';
+        }
+
+        return implode('|', $singleTypesAsString);
+    }
+
+    private function printUnionType(AttributeAwareUnionTypeNode $attributeAwareUnionTypeNode): string
+    {
+        $unionedTypes = [];
+
+        if ($attributeAwareUnionTypeNode->isWrappedWithBrackets()) {
+            return $attributeAwareUnionTypeNode . '[]';
+        }
+
+        foreach ($attributeAwareUnionTypeNode->types as $unionedType) {
+            $unionedTypes[] = $unionedType . '[]';
+        }
+
+        return implode('|', $unionedTypes);
```

Comment:
This was the reason for normalized, to avoid these custom printers for union/array mixes :)

Reply:
But if I understand it correctly... now we have to handle them in order to generate correct docblock. 

All this type related part of the Rector looks way too complicated :) a bit less complicated after all the investigations I made but anyway...

Reply:
Yes, by removing normalization on one place, now this part becomes more complicated :)